### ISSUE TITLE
[vDisk] Kali Linux 2020.3 starts booting and drops into initramfs

### ISSUE THREAD

**Issue by luminouw** [ventoy/Ventoy#535]
Hello there,

just discovered Ventoy and playing with it with some success so far. I wanted to try the boot from VDI feature for a Kali Linux machine but can't get it to boot completely.

My setup

- Ventoy 1.0.24
- BIOS : Legacy Mode
- VDI Fixed size
- Kali Linux installed on VDI disk in Legacy Mode (one partition with everything in it)

I downloaded and applied the `vtoyboot.sh` script from the ventoy/vtoyboot repo and it applied successfully.
I renamed the VDI file to `kali.vdi.vtoy`

When booting on my USB stick, Ventoy menu loads and lets me select the Kali VDI file.
From this point there are 2 possibilities:

1. Grub starts and immediatly stop with an error message saying the boot and root partition couldn't be found
From the Grub command line I can list disks and can only see the Ventoy USB drive and it's partitions but the VDI remains invisible.

OR

2. Grub starts and display the Kali boot menu (meaning it found Grub inside the VDI file and the /boot/grub/grub.cfg file). Selecting the first entry, the kernel starts to boot and the Kali logo is displaying. It then hangs and drops into an `initramfs` shell.
From this shell I can't see the VDI disk anywhere. My guess is the kernel starts to boot, and then somehow discovers the other disks (physical) from the system, which confuses the boot process as they do not contain any Kali installation.

I also noticed that scenario 1 often happens when doing a reboot with the usb key still plugged in, and scenario 2 is more likely to happen if I unplug completely the usb key before turning on my system.

Let me know how I could help you further to debug this issue as the boot from VDI features looks very interesting 👍

**Comment by ventoy** [comment]
1. You install kali by VirtualBox to a VDI file? The VDI file is originally at USB or the VDI file is firstly at local disk and you copy it to the USB drive after you installed kali in it?

2. You mean scenario 2 is that you unplug the USB drive and re-plug it again and then turn on your PC?

3. scenario 2 is most likely driver missing for your USB drive. You can boot a Kali ISO file in your PC and find the driver used for the USB drive after boot, then I can update the vtoyboot script to add this driver

4. Currently I have no idea about scenario 1

**Comment by luminouw** [comment]
> 1. You install kali by VirtualBox to a VDI file? The VDI file is originally at USB or the VDI file is firstly at local disk and you copy it to the USB drive after you installed kali in it?
Yes Kali is installed with Virtualbox onto a VDI disk file.
I tried both solutions: 
- copying the VDI file from local disk to usb drive + adding extension .vtoy
- using the .vdi.vtoy file stored on the usb drive directly in Virtualbox (works the same, saves some time because I don't need to copy the file each time I update the machine but it's a bit slower than my local disk due to the usb drive I am using. To be noted that this would be the situation I'd like to keep in the future).

> 2. You mean scenario 2 is that you unplug the USB drive and re-plug it again and then turn on your PC?
Yes, I update the .vdi.vtoy file (copy or direct modification, doesn't matter), then reboot the PC with the USB drive still attached. When I do it like this it is more likely to send me into direct error (scenario 1). If I unplug the usb drive, and plug it back during early boot process, it goes into scenario 2 (again maybe that's just a coincidence because I don't see why it would be any different).

> 3. scenario 2 is most likely driver missing for your USB drive. You can boot a Kali ISO file in your PC and find the driver used for the USB drive after boot, then I can update the vtoyboot script to add this driver
I am not sure that's a driver problem, when I'm booted into the `initramfs` shell, I can type and check the `dmesg` output and I can see the Ventoy USB drive being found and assigned a `/dev/sdX` name.

I don't know how the VDI file is supposed to be magically listed as `/dev/sdY` and booted from (is it the `vtoyboot` script that patches the boot process to do this ?), but I thought maybe during the boot process, the local disks from my machine are found and assigned `/dev` references and somehow replace something where the system expects the root partition in the VDI file (`/dev/sda1` in my case).

> 4. Currently I have no idea about scenario 1
Yeah me neither, I thought it had to do with the disk UUID so I gave it a go by changing `GRUB_DISABLE_LINUX_UUID` to `true` in my `/etc/default/grub` file and updating Grub after that (on my Kali system), as well as modifying the `/etc/fstab` file to use device names instead of UUIDs but when scenario 1 happens I get the same error.

> 5. You can test a distro in the list https://www.ventoy.net/en/plugin_vtoyboot.html  if possible
Will do with the Ubuntu Desktop image and get back to you : )

**Comment by ventoy** [comment]
Please retry with ventoy-1.0.25 and vtoyboot-1.0.1



    


            On 10/21/2020 21:09, luminouw wrote: 


You install kali by VirtualBox to a VDI file? The VDI file is originally at USB or the VDI file is firstly at local disk and you copy it to the USB drive after you installed kali in it?
Yes Kali is installed with Virtualbox onto a VDI disk file.
I tried both solutions:



copying the VDI file from local disk to usb drive + adding extension .vtoy
using the .vdi.vtoy file stored on the usb drive directly in Virtualbox (works the same, saves some time because I don't need to copy the file each time I update the machine but it's a bit slower than my local disk due to the usb drive I am using. To be noted that this would be the situation I'd like to keep in the future).



You mean scenario 2 is that you unplug the USB drive and re-plug it again and then turn on your PC?
Yes, I update the .vdi.vtoy file (copy or direct modification, doesn't matter), then reboot the PC with the USB drive still attached. When I do it like this it is more likely to send me into direct error (scenario 1). If I unplug the usb drive, and plug it back during early boot process, it goes into scenario 2 (again maybe that's just a coincidence because I don't see why it would be any different).




scenario 2 is most likely driver missing for your USB drive. You can boot a Kali ISO file in your PC and find the driver used for the USB drive after boot, then I can update the vtoyboot script to add this driver
I am not sure that's a driver problem, when I'm booted into the initramfs shell, I can type and check the dmesg output and I can see the Ventoy USB drive being found and assigned a /dev/sdX name.


I don't know how the VDI file is supposed to be magically listed as /dev/sdY and booted from (is it the vtoyboot script that patches the boot process to do this ?), but I thought maybe during the boot process, the local disks from my machine are found and assigned /dev references and somehow replace something where the system expects the root partition in the VDI file (/dev/sda1 in my case).


Currently I have no idea about scenario 1
Yeah me neither, I thought it had to do with the disk UUID so I gave it a go by changing GRUB_DISABLE_LINUX_UUID to true in my /etc/default/grub file and updating Grub after that (on my Kali system), as well as modifying the /etc/fstab file to use device names instead of UUIDs but when scenario 1 happens I get the same error.




You can test a distro in the list https://www.ventoy.net/en/plugin_vtoyboot.html  if possible
Will do with the Ubuntu Desktop image and get back to you : )



—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/ventoy/Ventoy/issues/535#issuecomment-713557071",
"url": "https://github.com/ventoy/Ventoy/issues/535#issuecomment-713557071",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by luminouw** [comment]
> Please retry with ventoy-1.0.25 and vtoyboot-1.0.1

Updated to 1.0.25 and reapplied the vtoyboot-1.0.1 without success, here are the screenshots from initramfs with dmesg output and error at boot.

The UUID mentionned in the error is the partition UUID corresponding to the boot / root partition where Kali Linux is installed (as seen when booted in VirtualBox).

![initram_dmesg](https://user-images.githubusercontent.com/410004/96778362-e4b15000-13eb-11eb-9136-606f54dc1a2c.jpg)
![initram_exit](https://user-images.githubusercontent.com/410004/96778364-e5e27d00-13eb-11eb-8a6b-e2518368cf4f.jpg)

**Comment by ventoy** [comment]
Please run 2 commands in the initramfs shell:
```
blkid
vtoydump -v
```

**Comment by luminouw** [comment]
There you go.

Kinda weird that the vtoydump binary cannot be found while there is a vtoy dump in it.

For the record I tried this with an Ubuntu 20.04 virtual machine as well (same setup, legacy mode, fixed vdi, ran the vtoyboot script as sudo inside the vm), and i end up with similar results.

![blkid](https://user-images.githubusercontent.com/410004/96912731-1802e600-14a3-11eb-9444-2f3cfd1da2c3.jpg)
![vtoydump](https://user-images.githubusercontent.com/410004/96912738-189b7c80-14a3-11eb-9297-fdf941971816.jpg)

From here I can try to manually execute all the step in the vtoyboot script or execute it with the `-x` flag to see where it fails.

**Comment by ventoy** [comment]
When you  run  vtoyboot.sh,  the initramfs file will be updated and a `vtoydump` binary will be added to the new initramfs file.
So I don't known why it's not exist in the initramfs, it's OK in my test.

You can make some debug for vtoyboot.sh 
When vtoyboot.sh finished, you can use  `lsinitramfs /boot/xxxx.img | grep vtoy` to check.

**Comment by luminouw** [comment]
Ok so after investigating why the vtoy binaries were not copied into the initramfs image, I figured that the hook script wasn't executable. So I performed a `chmod +x` on it and then using `lsinitramfs` I saw the binaries in the image.

Except it didn't work.

So I asked myself "Hey why was that file not executable ? Maybe they forgot to commit it as such in the Github repository ?" and looked around and re-read the documentation to see if I did miss something.

And the difference was that I was cloning the repository instead of getting the release files, and when cloning the repository, all files that should be executable weren't marked as such, even the binaries, so that's why it wasn't working even after I updated the executable bit on the hook script.

Long story short, I should pay more attention to the doc when it says to get the release files : ) 

Just a note here though, the releases files have a `tar.gz` extension but they aren't a `gzip` archive, just a classic `tar` (so you might want to change the extension or `gzip` the following releases).

It works now, sorry for the hassle ... Now on my way to try it in UEFI mode this time : )

**Comment by ventoy** [comment]
Good news!  I have updated `vtoybooy-1.0.2`  https://github.com/ventoy/vtoyboot/releases/
1. Fix the tar.gz file format issue
2. Add chmod for the shell script

So now it may OK even you clone the repo and run the script.

**Comment by luminouw** [comment]
Great reactivity, awesome !

Struggling now with UEFI mode :
- Keys imported into my Secure Boot
- Ventoy 1.0.25 with Secure Boot support enabled
- Using a Kali 2020.3 freshly installed in Virtualbox with UEFI enabled

When trying to boot the UEFI version, I can select the image but then I immediatly get
```
No bootfile found for UEFI!
Maybe the image does not support X64 UEFI!
```
I'm going to look why it fails at this stage and I'll open another issue if needed.

Great work still !

**Comment by ventoy** [comment]
Can you boot ISO file normally in UEFI mode with your computer?

**Comment by luminouw** [comment]
I just did some tests with the following files
- Kali VDI file installed in UEFI mode
- kali-linux-2020.3-installer-amd64.iso
- ubuntu-20.04.1-desktop-amd64.iso

With following settings : 
- UEFI boot enabled
- Secure Boot enabled
- ❌ Kali VDI file installed in UEFI mode
```
No bootfile found for UEFI!
Maybe the image does not support X64 UEFI!
```
- ❌ kali-linux-2020.3-installer-amd64.iso
It gets to the menu (ugly menu in text mode) where I can choose the type of install, if I choose "Graphic Install" I get the following message
```
error: Secure Boot forbids loading module from
(hd2)/boot/grub/x86_64-efi/linux.mod.
error: Secure Boot forbids loading module from
(hd2)/boot/grub/x86_64-efi/linux.mod.

Press any key to continue ...
```
- ✔️ ubuntu-20.04.1-desktop-amd64.iso
Presents the boot menu and choosing "Ubuntu" leads me straight to the "Try Ubuntu" button where the live environment works as expected.

Now if I **disable** Secure Boot and **keep** UEFI:
- ❌ Kali VDI file installed in UEFI mode, still same error
```
No bootfile found for UEFI!
Maybe the image does not support X64 UEFI!
```
- ✔️ kali-linux-2020.3-installer-amd64.iso. I am presented with a clean graphical menu where I can choose to install Kali, and if selected the installation process starts (no more Secure Boot error message).
- ✔️ ubuntu-20.04.1-desktop-amd64.iso, works the same.

**Comment by luminouw** [comment]
The EFI boot file on my Kali install seems to be located at `/boot/efi/EFI/kali/grubx64.efi`, not sure if that helps for Ventoy to find it.

**Comment by luminouw** [comment]
More info from `efibootmgr` : 

```
luminouw@kali:~$ sudo efibootmgr -v                                                                                                                                                                                                   
BootCurrent: 0004                                                                                                                                                                                                                          
Timeout: 0 seconds
BootOrder: 0004,0000,0001,0002,0003
Boot0000* UiApp FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(462caa21-7614-4503-836e-8ab6f4662331)
Boot0001* UEFI VBOX CD-ROM VB2-01700376         PciRoot(0x0)/Pci(0x1,0x1)/Ata(1,0,0)N.....YM....R,Y.
Boot0002* UEFI VBOX HARDDISK VB8a711cf7-8405e1c6        PciRoot(0x0)/Pci(0xd,0x0)/Sata(0,65535,0)N.....YM....R,Y.
Boot0003* EFI Internal Shell    FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(7c04a583-9e3e-4f1c-ad65-e05268d0b4d1)
Boot0004* kali  HD(1,GPT,1caabad7-aaa4-433f-a36a-cd9783f968ae,0x800,0x100000)/File(\EFI\kali\grubx64.efi)
luminouw@aldebaran:~$ sudo efibootmgr -v                                                                                                                                                                                                   
BootCurrent: 0004                                                                                                                                                                                                                          
Timeout: 0 seconds
BootOrder: 0004,0000,0001,0002,0003
Boot0000* UiApp FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(462caa21-7614-4503-836e-8ab6f4662331)
Boot0001* UEFI VBOX CD-ROM VB2-01700376         PciRoot(0x0)/Pci(0x1,0x1)/Ata(1,0,0)N.....YM....R,Y.
Boot0002* UEFI VBOX HARDDISK VB8a711cf7-8405e1c6        PciRoot(0x0)/Pci(0xd,0x0)/Sata(0,65535,0)N.....YM....R,Y.
Boot0003* EFI Internal Shell    FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(7c04a583-9e3e-4f1c-ad65-e05268d0b4d1)
Boot0004* kali  HD(1,GPT,1caabad7-aaa4-433f-a36a-cd9783f968ae,0x800,0x100000)/File(\EFI\kali\grubx64.efi)
```

**Comment by ventoy** [comment]
Ventoy only load `EFI\boot\bootx64.efi `  and normally there should be a `/boot/efi/EFI/boot/bootx64.efi` file after you fresh install kali.  
You can make a copy `/boot/efi/EFI/kali/grubx64.efi`  to  `/boot/efi/EFI/boot/bootx64.efi` in the VDI system and retry.

**Comment by luminouw** [comment]
Copying the file in the expected location did the trick.

:x: With Secure Boot **enabled** I get the following ```error: Secure Boot forbids loading module from (hd2,gpt2)/boot/grub/x86_64-efi/normal.mod.```

✔️ With Secure Boot **disabled**, it boots up and loads the system perfectly : )

Got to figure how to add my install as trusted now and I'll be all good !

Thanks for the support : )
